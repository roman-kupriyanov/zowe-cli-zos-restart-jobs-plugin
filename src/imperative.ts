/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { IImperativeConfig } from "@zowe/imperative";

const config: IImperativeConfig = {
    commandModuleGlobs: ["**/cli/*/*.definition!(.d).*s"],
    pluginHealthCheck: __dirname + "/healthCheck.Handler",
    pluginSummary: "Zowe CLI restart z/OS jobs plug-in",
    rootCommandDescription: "Plugin, which allows to restart z/OS jobs",
    productDisplayName: "Zowe CLI z/OS Jobs Restart Plug-in",
    name: "zos-restart-jobs"
};

export = config;
