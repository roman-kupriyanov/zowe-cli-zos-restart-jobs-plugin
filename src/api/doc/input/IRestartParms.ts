/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { ITaskWithStatus } from "@zowe/imperative";

/**
 * Interface for restart job API
 * @export
 * @interface IRestartParms
 */
export interface IRestartParms {

    /**
     * Returns spool content if this option used
     */
    viewAllSpoolContent?: boolean;

    /**
     * Wait for the job to reach output status
     */
    waitForActive?: boolean;


    /**
     * Wait for the job to reach output status
     */
    waitForOutput?: boolean;

    /**
     * Local directory path to download output of the job
     */
    directory?: string;

    /**
     * A file extension to save the job output with
     */
    extension?: string;

    /**
     * Task status object used by CLI handlers to create progress bars
     * for certain job restart requests
     * Optional
     */
    task?: ITaskWithStatus;

}
