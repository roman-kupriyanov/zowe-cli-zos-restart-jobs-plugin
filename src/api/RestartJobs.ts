/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { IJob, GetJobs, SubmitJobs, JOB_STATUS } from "@zowe/cli";
import { AbstractSession, ImperativeError, ImperativeExpect } from "@zowe/imperative";
import { IRestartParms } from "./doc/input/IRestartParms";

/**
 * Class of restart jobs APIs for usage within the CLI and programmatically from node scripts
 * @export
 * @class RestartJobs
 */
export class RestartJobs {

    /**
     * Get JCL for job and modify it for a restart
     * @static
     * @param {AbstractSession} session - z/OSMF connection info
     * @param {string} stepname - name of a step, which to restart from
     * @param {IJob} job - job to get JCL for
     * @throws {ImperativeError} - throws an error if specified step name is not found in JCL
     * @returns {Promise<string>} - promise that resolves to a string, which contains JCL modified for a restart
     * @memberof RestartJobs
     */
    public static async getRestartJclForJob(session: AbstractSession, stepname: string, job: IJob) {
        const jobJcl: string = await GetJobs.getJclForJob(session, job);

        const newJclLines: string[] = [];
        const restartParamLine: string = `// RESTART=(${stepname.toUpperCase()})`;
        let isStepFound: boolean = false;
        let isRestartParmFound: boolean = false;

        for (const line of jobJcl.split("\n")) {

            // Transform to upper case to check for substrings
            const upperCasedLine = line.toUpperCase();

            // Line actually to be added in new JCL
            let lineToAdd: string = line.toString();

            // Do not try to process comment lines
            if (!upperCasedLine.startsWith("//*")) {

                if (upperCasedLine.indexOf("JOB") >= 0) {
                    // Remove redundant `jobid` at the end of JOB statement
                    lineToAdd = line.replace(job.jobid, "");
                }

                if (!isRestartParmFound) {
                    // Trim white spaces just in case to be able to check last line symbol
                    lineToAdd = lineToAdd.trim();

                    // If already specified RESTART= parm is found -> replace it with new step name
                    // and stop searching
                    if (lineToAdd.match(/RESTART=\(\S*\)/)) {
                        lineToAdd = lineToAdd.replace(/RESTART=\(\S*\)/, `RESTART=(${stepname})`);
                        isRestartParmFound = true;
                    }

                    // If no RESTART= parm is found inline and it is no continuation to next line ->
                    // add continuation and RESTART= parm on next line, then stop searching
                    if (!isRestartParmFound && !lineToAdd.endsWith(",")) {
                        lineToAdd += ",";
                        newJclLines.push(lineToAdd);
                        newJclLines.push(restartParamLine);
                        isRestartParmFound = true;
                        continue;
                    }
                }

                // Check if specified step name really exists in JCL
                if (upperCasedLine.indexOf("EXEC") >= 0) {
                    if (upperCasedLine.startsWith(`//${stepname.toUpperCase()}`)) {
                        isStepFound = true;
                    }
                }
            }

            newJclLines.push(lineToAdd);

        }

        if (!isStepFound) {
            throw new ImperativeError({
                msg: `Step name ${stepname} is not found in a job with jobid ${job.jobid}`
            });
        }

        return newJclLines.join("\n");

    }

    /**
     * Check if job is failed and return its JCL prepared for restart
     * @static
     * @param {AbstractSession} session - z/OSMF connection info
     * @param {string} jobid - job id to be translated into parms object
     * @param {string} stepname - a name of a step, which to restart from
     * @throws {ImperativeError} - throws an error if job is in ACTIVE state and return code is not "CC 0000"
     * @returns {Promise<IJob>} - Promise that resolves to an IJob document with details about the restarted job
     * @memberof RestartJobs
     */
    public static async getFailedJobRestartJcl(session: AbstractSession, jobid: string, stepname: string) {

        // Get the job details
        const job: IJob = await GetJobs.getJob(session, jobid);

        const errorMessagePrefix: string =
            `Restarting job with id ${jobid} on ${session.ISession.hostname}:${session.ISession.port} failed: `;

        ImperativeExpect.toBeEqual(job.status, JOB_STATUS.OUTPUT,
                                      errorMessagePrefix + "Job status is ACTIVE, OUTPUT is required");
        ImperativeExpect.toNotBeEqual(job.retcode, "CC 0000",
                                      errorMessagePrefix + "Job status is successful, failed is required");

        // Get the restart job JCL
        return this.getRestartJclForJob(session, stepname, job);

    }

    /**
     * Restart a job from a specific step
     * @static
     * @param {AbstractSession} session - z/OSMF connection info
     * @param {string} jobid - job id to be translated into parms object
     * @param {string} stepname - name of a step, which to restart from
     * @throws {ImperativeError} - throws an error if job is in ACTIVE state and return code is not "CC 0000"
     * @returns {Promise<IJob>} - promise that resolves to an IJob document with details about the restarted job
     * @memberof RestartJobs
     */
    public static async restartFailedJob(session: AbstractSession, jobid: string, stepname: string) {

        // Get the restart job JCL
        const restartJobJcl: string = await this.getFailedJobRestartJcl(session, jobid, stepname);

        // Re-submit restart job JCL
        return SubmitJobs.submitJcl(session, restartJobJcl);

    }

    /**
     * Restart a job from a specific step
     * @static
     * @param {AbstractSession} session - z/OSMF connection info
     * @param {string} jobid - job id to be translated into parms object
     * @param {string} stepname - name of a step, which to restart from
     * @param {IRestartParms} parms - special object with restart parameters (see for details)
     * @throws {ImperativeError} - throws an error if job is in ACTIVE state and return code is not "CC 0000"
     * @returns {Promise<IJob | ISpoolFile[]>} - promise that resolves to an IJob document with details about the restarted job or
     *                                           into a list of ISpoolFile documents with spool data set content
     * @memberof RestartJobs
     */
    public static async restartFailedJobWithParms(session: AbstractSession, jobid: string, stepname: string,
                                                  parms: IRestartParms) {

        // Get the restart job JCL
        const restartJobJcl: string = await this.getFailedJobRestartJcl(session, jobid, stepname);

        // Transform into ISubmitParms structure
        const submitParms = {
            jclSource: undefined as any,
            viewAllSpoolContent: parms.viewAllSpoolContent,
            directory: parms.directory,
            extension: parms.extension,
            waitForActive: parms.waitForActive,
            waitForOutput: parms.waitForOutput,
            task: parms.task
        };

        // Re-submit restart job JCL
        return SubmitJobs.submitJclString(session, restartJobJcl, submitParms);

    }

}
