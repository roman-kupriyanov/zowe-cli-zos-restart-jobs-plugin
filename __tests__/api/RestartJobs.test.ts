/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { GetJobs, SubmitJobs, IJob } from "@zowe/cli";
import { Session, TaskProgress, TaskStage } from "@zowe/imperative";
import { RestartJobs } from "../../src/api/RestartJobs";
import { IRestartParms } from "../../src/api/doc/input/IRestartParms";

describe("RestartJobs tests", () => {

    const jobid: string = "JOB04541";
    const stepname: string = "STEP02";

    const job: IJob = {
        owner: "Z12345",
        phase: 20,
        subsystem: "JES2",
        "phase-name": "Job is on the hard copy queue",
        "job-correlator": "J0004541SVSCJES2D74CB05A.......:",
        type: "JOB",
        url: "https://127.0.0.1:443/zosmf/restjobs/jobs/J0004541SVSCJES2D74CB05A.......%3A",
        jobid,
        class: "A",
        "files-url": "https://127.0.0.1:443/zosmf/restjobs/jobs/J0004541SVSCJES2D74CB05A.......%3A/files",
        jobname: "TESTJOB",
        status: "OUTPUT",
        retcode: "ABEND S806"
    };

    const restartJobJcl: string = "//TESTJOB JOB (ACCTINFO),'user',CLASS=A,\n" +
                                  "//             RESTART=(STEP02),\n" +
                                  "//STEP01   EXEC PGM=XYZ\n" +
                                  "//STEP02   EXEC PGM=IEFBR14";


    const restartJob: IJob = {
        owner: "",
        phase: 128,
        subsystem: "JES2",
        "phase-name": "Job is active in input processing",
        "job-correlator": "J0004565SVSCJES2D74D1C38.......:",
        type: "JOB",
        url: "https://127.0.0.1:443/zosmf/restjobs/jobs/J0004565SVSCJES2D74D1C38.......%3A",
        jobid,
        class: "A",
        "files-url": "https://127.0.0.1:443/zosmf/restjobs/jobs/J0004565SVSCJES2D74D1C38.......%3A/files",
        jobname: "TESTJOB",
        status: "INPUT",
        retcode: null
    };

    const dummySession = new Session({ hostname: "dummy" });

    describe("getRestartJclForJob tests", () => {

        const getJclForJobSpy = jest.spyOn(GetJobs, "getJclForJob");

        beforeEach(() => {

            getJclForJobSpy.mockClear();

        });

        it("should success with single line JOB statement", async () => {

            const jobJcl: string = "//TESTJOB JOB (ACCTINFO),'user',CLASS=A                                JOB04541\n" +
                                   "//STEP01   EXEC PGM=XYZ\n" +
                                   "//STEP02   EXEC PGM=IEFBR14";

            const modifiedJobJcl: string = "//TESTJOB JOB (ACCTINFO),'user',CLASS=A,\n" +
                                           "// RESTART=(STEP02)\n" +
                                           "//STEP01   EXEC PGM=XYZ\n" +
                                           "//STEP02   EXEC PGM=IEFBR14";

            getJclForJobSpy.mockImplementation(() => Promise.resolve(jobJcl));

            const resultJobJcl: string = await RestartJobs.getRestartJclForJob(dummySession, stepname, job);

            expect(getJclForJobSpy).toHaveBeenCalledWith(dummySession, job);
            expect(resultJobJcl).toEqual(modifiedJobJcl);
        });

        it("should success with multi line JOB statement", async () => {

            const jobJcl: string = "//TESTJOB JOB (ACCTINFO),'user',                                       JOB04541\n" +
                                   "//             CLASS=A\n" +
                                   "//STEP01   EXEC PGM=XYZ\n" +
                                   "//STEP02   EXEC PGM=IEFBR14";

            const modifiedJobJcl: string = "//TESTJOB JOB (ACCTINFO),'user',\n" +
                                           "//             CLASS=A,\n" +
                                           "// RESTART=(STEP02)\n" +
                                           "//STEP01   EXEC PGM=XYZ\n" +
                                           "//STEP02   EXEC PGM=IEFBR14";

            getJclForJobSpy.mockImplementation(() => Promise.resolve(jobJcl));

            const resultJobJcl: string = await RestartJobs.getRestartJclForJob(dummySession, stepname, job as IJob);

            expect(getJclForJobSpy).toHaveBeenCalledWith(dummySession, job);
            expect(resultJobJcl).toEqual(modifiedJobJcl);
        });

        it("should success with single line JOB statement and RESTART= already specified there", async () => {

            const jobJcl: string = "//TESTJOB JOB (ACCTINFO),'user',RESTART=(STEP01)                       JOB04541\n" +
                                   "//STEP01   EXEC PGM=XYZ\n" +
                                   "//STEP02   EXEC PGM=IEFBR14";

            const modifiedJobJcl: string = "//TESTJOB JOB (ACCTINFO),'user',RESTART=(STEP02)\n" +
                                           "//STEP01   EXEC PGM=XYZ\n" +
                                           "//STEP02   EXEC PGM=IEFBR14";

            getJclForJobSpy.mockImplementation(() => Promise.resolve(jobJcl));

            const resultJobJcl: string = await RestartJobs.getRestartJclForJob(dummySession, stepname, job as IJob);

            expect(getJclForJobSpy).toHaveBeenCalledWith(dummySession, job);
            expect(resultJobJcl).toEqual(modifiedJobJcl);
        });

        it("should success with multi line JOB statement and RESTART= already specified there", async () => {

            const jobJcl: string = "//TESTJOB JOB (ACCTINFO),'user',RESTART=(STEP01),                      JOB04541\n" +
                                   "//             CLASS=A\n" +
                                   "//STEP01   EXEC PGM=XYZ\n" +
                                   "//STEP02   EXEC PGM=IEFBR14";

            const modifiedJobJcl: string = "//TESTJOB JOB (ACCTINFO),'user',RESTART=(STEP02),\n" +
                                           "//             CLASS=A\n" +
                                           "//STEP01   EXEC PGM=XYZ\n" +
                                           "//STEP02   EXEC PGM=IEFBR14";

            getJclForJobSpy.mockImplementation(() => Promise.resolve(jobJcl));

            const resultJobJcl: string = await RestartJobs.getRestartJclForJob(dummySession, stepname, job as IJob);

            expect(getJclForJobSpy).toHaveBeenCalledWith(dummySession, job);
            expect(resultJobJcl).toEqual(modifiedJobJcl);
        });

        it("should success with multi line JOB statement and RESTART= already specified on other line", async () => {

            const jobJcl: string = "//TESTJOB JOB (ACCTINFO),'user',                                       JOB04541\n" +
                                   "//             RESTART=(STEP01),\n" +
                                   "//             CLASS=A\n" +
                                   "//STEP01   EXEC PGM=XYZ\n" +
                                   "//STEP02   EXEC PGM=IEFBR14";

            const modifiedJobJcl: string = "//TESTJOB JOB (ACCTINFO),'user',\n" +
                                           "//             RESTART=(STEP02),\n" +
                                           "//             CLASS=A\n" +
                                           "//STEP01   EXEC PGM=XYZ\n" +
                                           "//STEP02   EXEC PGM=IEFBR14";

            getJclForJobSpy.mockImplementation(() => Promise.resolve(jobJcl));

            const resultJobJcl: string = await RestartJobs.getRestartJclForJob(dummySession, stepname, job as IJob);

            expect(getJclForJobSpy).toHaveBeenCalledWith(dummySession, job);
            expect(resultJobJcl).toEqual(modifiedJobJcl);
        });

        it("should success with multi line JOB statement and RESTART= already specified on other line with extra param", async () => {

            const jobJcl: string = "//TESTJOB JOB (ACCTINFO),'user',                                       JOB04541\n" +
                                   "//             CLASS=A,RESTART=(STEP01),\n" +
                                   "//             CLASS=A\n" +
                                   "//STEP01   EXEC PGM=XYZ\n" +
                                   "//STEP02   EXEC PGM=IEFBR14";

            const modifiedJobJcl: string = "//TESTJOB JOB (ACCTINFO),'user',\n" +
                                           "//             CLASS=A,RESTART=(STEP02),\n" +
                                           "//             CLASS=A\n" +
                                           "//STEP01   EXEC PGM=XYZ\n" +
                                           "//STEP02   EXEC PGM=IEFBR14";

            getJclForJobSpy.mockImplementation(() => Promise.resolve(jobJcl));

            const resultJobJcl: string = await RestartJobs.getRestartJclForJob(dummySession, stepname, job as IJob);

            expect(getJclForJobSpy).toHaveBeenCalledWith(dummySession, job);
            expect(resultJobJcl).toEqual(modifiedJobJcl);
        });

        it("should success with multi line JOB statement and RESTART= already specified on other line with extra params", async () => {

            const jobJcl: string = "//TESTJOB JOB (ACCTINFO),'user',                                       JOB04541\n" +
                                   "//             CLASS=A,RESTART=(STEP01),CLASS=A,\n" +
                                   "//             CLASS=A\n" +
                                   "//STEP01   EXEC PGM=XYZ\n" +
                                   "//STEP02   EXEC PGM=IEFBR14";

            const modifiedJobJcl: string = "//TESTJOB JOB (ACCTINFO),'user',\n" +
                                           "//             CLASS=A,RESTART=(STEP02),CLASS=A,\n" +
                                           "//             CLASS=A\n" +
                                           "//STEP01   EXEC PGM=XYZ\n" +
                                           "//STEP02   EXEC PGM=IEFBR14";

            getJclForJobSpy.mockImplementation(() => Promise.resolve(jobJcl));

            const resultJobJcl: string = await RestartJobs.getRestartJclForJob(dummySession, stepname, job as IJob);

            expect(getJclForJobSpy).toHaveBeenCalledWith(dummySession, job);
            expect(resultJobJcl).toEqual(modifiedJobJcl);
        });

        it("should error when no step name found in jcl", async () => {

            const jobJcl: string = "//TESTJOB JOB (ACCTINFO),'user',CLASS=A                                JOB04541\n" +
                                   "//STEP01   EXEC PGM=XYZ\n" +
                                   "//STEP02   EXEC PGM=IEFBR14";

            getJclForJobSpy.mockImplementation(() => Promise.resolve(jobJcl));

            try {
                await RestartJobs.getRestartJclForJob(dummySession, "STEP03", job as IJob);
            }
            catch (e) {
                expect(e).toMatchSnapshot();
            }

            expect(getJclForJobSpy).toHaveBeenCalledWith(dummySession, job);

        });

    });

    describe("getFailedJobRestartJcl tests", () => {

        const getJobSpy = jest.spyOn(GetJobs, "getJob");
        const getRestartJclForJobSpy = jest.spyOn(RestartJobs, "getRestartJclForJob");

        beforeEach(() => {

            getJobSpy.mockClear();
            getRestartJclForJobSpy.mockClear();

        });

        it("should success", async () => {

            getJobSpy.mockImplementation(() => Promise.resolve(job));
            getRestartJclForJobSpy.mockImplementation(() => Promise.resolve(restartJobJcl));

            const resultJobJcl: string = await RestartJobs.getFailedJobRestartJcl(dummySession, jobid, stepname);

            expect(getJobSpy).toHaveBeenCalledWith(dummySession, jobid);
            expect(getRestartJclForJobSpy).toHaveBeenCalledWith(dummySession, stepname, job);
            expect(resultJobJcl).toEqual(restartJobJcl);

        });

    });

    describe("restartFailedJob tests", () => {

        const getFailedJobRestartJclSpy = jest.spyOn(RestartJobs, "getFailedJobRestartJcl");
        const submitJclSpy = jest.spyOn(SubmitJobs, "submitJcl");

        beforeEach(() => {

            getFailedJobRestartJclSpy.mockClear();
            submitJclSpy.mockClear();

        });

        it("should success", async () => {

            getFailedJobRestartJclSpy.mockImplementation(() => Promise.resolve(restartJobJcl));
            submitJclSpy.mockImplementation(() => Promise.resolve(restartJob));

            const resultJob = await RestartJobs.restartFailedJob(dummySession, jobid, stepname);

            expect(getFailedJobRestartJclSpy).toHaveBeenCalledWith(dummySession, jobid, stepname);
            expect(submitJclSpy).toHaveBeenCalledWith(dummySession, restartJobJcl);
            expect(resultJob).toEqual(restartJob);

        });

    });

    describe("restartFailedJobWithParms tests", () => {

        const getFailedJobRestartJclSpy = jest.spyOn(RestartJobs, "getFailedJobRestartJcl");
        const submitJclStringSpy = jest.spyOn(SubmitJobs, "submitJclString");

        beforeEach(() => {

            getFailedJobRestartJclSpy.mockClear();
            submitJclStringSpy.mockClear();

        });

        it("should success with no wait for output", async () => {

            const restartParms: IRestartParms = {
                viewAllSpoolContent: false,
                directory: "",
                extension: ".txt",
                waitForActive: false,
                waitForOutput: false,
                task: {
                    statusMessage: "Restarting job",
                    percentComplete: TaskProgress.TEN_PERCENT,
                    stageName: TaskStage.IN_PROGRESS
                }
            };

            const submitParms = {
                jclSource: undefined as any,
                viewAllSpoolContent: restartParms.viewAllSpoolContent,
                directory: restartParms.directory,
                extension: restartParms.extension,
                waitForActive: restartParms.waitForActive,
                waitForOutput: restartParms.waitForOutput,
                task: restartParms.task
            }

            getFailedJobRestartJclSpy.mockImplementation(() => Promise.resolve(restartJobJcl));
            submitJclStringSpy.mockImplementation(() => Promise.resolve(restartJob));

            const resultJob = await RestartJobs.restartFailedJobWithParms(dummySession, jobid, stepname, restartParms);

            expect(getFailedJobRestartJclSpy).toHaveBeenCalledWith(dummySession, jobid, stepname);
            expect(submitJclStringSpy).toHaveBeenCalledWith(dummySession, restartJobJcl, submitParms);
            expect(resultJob).toEqual(restartJob);

        });

    });

});
