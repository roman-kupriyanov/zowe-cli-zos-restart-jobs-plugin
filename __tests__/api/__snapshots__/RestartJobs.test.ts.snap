// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RestartJobs tests getRestartJclForJob tests should error when no step name found in jcl 1`] = `[Error: Step name STEP03 is not found in a job with jobid JOB04541]`;
